File: Sources/AINotebookCore/UpdateCheck.swift
import Foundation

/// One GitHub release asset (decoded from the REST API shape).
public struct UpdateReleaseAsset: Codable, Equatable, Sendable {
    public let name: String
    public let browserDownloadUrl: String

    enum CodingKeys: String, CodingKey {
        case name
        case browserDownloadUrl = "browser_download_url"
    }

    public init(name: String, browserDownloadUrl: String) {
        self.name = name
        self.browserDownloadUrl = browserDownloadUrl
    }
}

/// One GitHub release (only the fields the update check needs).
public struct UpdateRelease: Codable, Equatable, Sendable {
    public let tagName: String
    public let prerelease: Bool
    public let htmlUrl: String
    public let assets: [UpdateReleaseAsset]

    enum CodingKeys: String, CodingKey {
        case tagName = "tag_name"
        case prerelease
        case htmlUrl = "html_url"
        case assets
    }

    public init(tagName: String, prerelease: Bool, htmlUrl: String, assets: [UpdateReleaseAsset]) {
        self.tagName = tagName
        self.prerelease = prerelease
        self.htmlUrl = htmlUrl
        self.assets = assets
    }
}

/// Result of an update evaluation.
public struct UpdateInfo: Equatable, Sendable {
    public let isUpdateAvailable: Bool
    public let latestVersion: String
    public let downloadURL: String
    public let releaseNotesURL: String

    public init(isUpdateAvailable: Bool, latestVersion: String, downloadURL: String, releaseNotesURL: String) {
        self.isUpdateAvailable = isUpdateAvailable
        self.latestVersion = latestVersion
        self.downloadURL = downloadURL
        self.releaseNotesURL = releaseNotesURL
    }

    public static let none = UpdateInfo(
        isUpdateAvailable: false, latestVersion: "", downloadURL: "", releaseNotesURL: ""
    )
}

/// Pure release-picking logic — no networking in this file (CI grep gate).
/// The fetch layer lives in the App target (`UpdateService`).
public enum UpdateCheck {
    public static let macAssetSuffix = "-macos.dmg"

    /// Picks the highest-semver non-prerelease release that carries an asset
    /// with `assetSuffix`; available iff strictly newer than `currentVersion`.
    public static func evaluate(
        releases: [UpdateRelease],
        currentVersion: String,
        assetSuffix: String
    ) -> UpdateInfo {
        guard let current = semverComponents(of: currentVersion) else { return .none }

        var best: (version: [Int], display: String, asset: UpdateReleaseAsset, notes: String)?
        for release in releases where !release.prerelease {
            guard let version = semverComponents(of: release.tagName),
                  let asset = release.assets.first(where: { $0.name.hasSuffix(assetSuffix) })
            else { continue }
            if best == nil || isGreater(version, than: best!.version) {
                best = (version, version.map(String.init).joined(separator: "."), asset, release.htmlUrl)
            }
        }

        guard let best, isGreater(best.version, than: current) else { return .none }
        return UpdateInfo(
            isUpdateAvailable: true,
            latestVersion: best.display,
            downloadURL: best.asset.browserDownloadUrl,
            releaseNotesURL: best.notes
        )
    }

    /// "v0.9.2" / "win-v0.8.0" / "0.9.2" → [major, minor, patch]; nil otherwise.
    static func semverComponents(of tag: String) -> [Int]? {
        var s = tag
        if s.hasPrefix("win-v") { s.removeFirst("win-v".count) }
        else if s.hasPrefix("v") { s.removeFirst(1) }
        let parts = s.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 3 else { return nil }
        var numbers: [Int] = []
        for part in parts {
            guard let n = Int(part), n >= 0 else { return nil }
            numbers.append(n)
        }
        return numbers
    }

    static func isGreater(_ a: [Int], than b: [Int]) -> Bool {
        for (x, y) in zip(a, b) {
            if x != y { return x > y }
        }
        return false
    }
}

File: windows/src/AINotebook.Core/UpdateCheck.cs
namespace AINotebook.Core;

/// One GitHub release asset (REST API shape; JSON mapping lives in UpdateChecker).
public sealed record UpdateReleaseAsset(string Name, string BrowserDownloadUrl);

/// One GitHub release — only the fields the update check needs.
public sealed record UpdateRelease(
    string TagName, bool Prerelease, string HtmlUrl, IReadOnlyList<UpdateReleaseAsset> Assets);

/// Result of an update evaluation.
public sealed record UpdateInfo(
    bool IsUpdateAvailable, string LatestVersion, string DownloadUrl, string ReleaseNotesUrl)
{
    public static readonly UpdateInfo None = new(false, "", "", "");
}

/// Pure release-picking logic; the fetch layer is UpdateChecker.
public static class UpdateCheck
{
    public const string WindowsAssetSuffix = "-windows-setup.exe";

    /// Picks the highest-semver non-prerelease release carrying an asset with
    /// `assetSuffix`; available iff strictly newer than `currentVersion`.
    public static UpdateInfo Evaluate(
        IReadOnlyList<UpdateRelease> releases, string currentVersion, string assetSuffix)
    {
        var current = SemverComponents(currentVersion);
        if (current is null) return UpdateInfo.None;

        int[]? bestVersion = null;
        UpdateReleaseAsset? bestAsset = null;
        string bestNotes = "";
        foreach (var release in releases)
        {
            if (release.Prerelease) continue;
            var version = SemverComponents(release.TagName);
            if (version is null) continue;
            UpdateReleaseAsset? asset = null;
            foreach (var a in release.Assets)
            {
                if (a.Name.EndsWith(assetSuffix, StringComparison.Ordinal)) { asset = a; break; }
            }
            if (asset is null) continue;
            if (bestVersion is null || IsGreater(version, bestVersion))
            {
                bestVersion = version;
                bestAsset = asset;
                bestNotes = release.HtmlUrl;
            }
        }

        if (bestVersion is null || bestAsset is null || !IsGreater(bestVersion, current))
            return UpdateInfo.None;

        return new UpdateInfo(
            true, string.Join('.', bestVersion), bestAsset.BrowserDownloadUrl, bestNotes);
    }

    /// "v0.9.2" / "win-v0.8.0" / "0.9.2" → [major, minor, patch]; null otherwise.
    internal static int[]? SemverComponents(string tag)
    {
        var s = tag;
        if (s.StartsWith("win-v", StringComparison.Ordinal)) s = s["win-v".Length..];
        else if (s.StartsWith('v')) s = s[1..];
        var parts = s.Split('.');
        if (parts.Length != 3) return null;
        var numbers = new int[3];
        for (var i = 0; i < 3; i++)
        {
            if (!int.TryParse(parts[i], out var n) || n < 0) return null;
            numbers[i] = n;
        }
        return numbers;
    }

    internal static bool IsGreater(int[] a, int[] b)
    {
        for (var i = 0; i < Math.Min(a.Length, b.Length); i++)
        {
            if (a[i] != b[i]) return a[i] > b[i];
        }
        return false;
    }
}
